DOCS-5000 Java Module update (#371)
mulesoft/docs-connectors: modules/ROOT/pages/java/java-argument-transformation.adoc

= Argument Transformation - Mule 4
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> Since Java Module version 1.1, this Module transforms the arguments 

Reviewer: ```suggestion
Java module 1.1 and later transforms the arguments 
```

Reviewer: I have a question out to the PMMs about the capitalization of Module and Component as part of these element names, because they are not productized.
My preference is lowercase, but I will not go through and change all of the names in this PR until I have gotten an answer about this. Just a heads up.

Reviewer: ADDENDUM: As suspected, the PMM advice (David Wang) is lowercase:

Kim Bostrom:house_with_garden:  1 hour ago
Our documentation refers to named modules (such as the Java module/Java Module) and named components of modules.
Do you have any guidance on how to capitalize these names? My default position is to cap. the identifying term but not "module" or "component."
But if these are proprietary to MuleSoft, then I would capitalize those words...which our legacy doc. leans toward.
Advice?

David Wang  6 minutes ago
These are generic things, so Java module is right.

given to a Java invocation even if they don't exactly
match the type expected by the Java method. The Java Module 
infers how to transform a parameter to what a method expects.

> 

Reviewer: If we are saying that the module "transforms" the arguments, is it accurate to say "even if they don't exactly match"? Isn't the transformation BECAUSE they don't exactly match? The second sentence implies this, in which case maybe this text should say that when a Java invocation contains arguments that do not match those expected by the Java method, the Java module transforms them to match those expectations.

Reviewer: As the introduction to a set of topics, I would also like to see the next-level topics introduced to the audience here, providing an expectation of what information, exactly, is in this topic set:

"It is useful to understand when arguments are transformed; to see an example of a simple transformation; to understand custom POJOs, collections, and maps; ....

This is a really good way to see if you have included too much in one topic set, and to doublecheck your topic organization and hierarchy.

> == When Are Arguments Transformed?

Reviewer: Avoid using titles that are complete sentences, including questions (unless working in an FAQ).

For each specified parameter, the Java Module checks whether its type matches the type

> expected by the method for that parameter or not. If it is not what the method expects, the

Reviewer: ```suggestion
expected by the method for that parameter. If not, the
```

Reviewer: This text does not really answer the question of "when" (which I thought implied time) but rather "under what circumstances." 
That said, this topic does not serve any purpose, because it repeats what is already conveyed in the topic set introduction. Suggest omitting.

Java Module transforms it to what is expected.

== Simple Example of Argument Transformation

> This is a simple example where one of the arguments is transformed.

Reviewer: ```suggestion
Following is a simple example in which one of the arguments given to a Java invocation is transformed.
```

Reviewer: ...or any other wording that specifies which arguments....

> The static method `min` from the `Math` class is called, but the class expects

Reviewer: ```suggestion
Assume that the static method `min` from the `Math` class is called, but the class expects
```

> to receive two `int` parameters. The call is made with a Number and a String:

Reviewer: ```suggestion
to receive two `int` parameters. The call is made with a number and a string:
```
This is to refer to the call elements as types. If you want to quote the CLI or the UI, replace with "...a Number value and a String value:"  GLOBAL in all writing: lowercase types of things unless making a point of representing the UI verbatim, in which case you can use the syntax or label or other name AS a name, following it with a common noun like "value."

</flow>
----

> In this case `arg1` would have been passed as a string, but it is transformed into an integer

Reviewer: ```suggestion
Without the Java module, `arg1` would have been passed as a string, but the module transforms it into an integer
```

before calling the Java method.

== Pass a Custom POJO to a Module as a Map

> Previously, when working with custom POJOs, users explicitly declared in the DataWeave

Reviewer: Can you be more specific about "previously" by replacing it with a product and version number or some other specific milestone?

Reviewer: Suggest spelling out POJO and then presenting it in parens.
QUESTION: 
Does "custom" clash with the idea of "plain old" or ordinary Java object?

expression what type the data to transform. Now, the Java Module automatically

> recognizes what is the expected type for a given argument and tries to transform the given

Reviewer: ```suggestion
recognizes the expected type for a given argument and tries to transform it.
```

> data to the expected type. 

Reviewer: OMIT

Most of my suggested rewording and omissions are an effort to reduce repetition/verbosity.

> In the examples that follow, a Map parameter is transformed

Reviewer: More than one example of this scenario?

Reviewer: Also, since this is really a task, with steps, is it more precise to say that "you" configure and invoke a process in which the Java module transforms a Map parameter into a custom POJO?

            #[{
            'car': {
                'doors' : "Four doors",
                'engine' : "120 Horsepower engine",
                'wheels' : "17 Inch chromed rims"
            }
            }]
        </java:args>
    </java:invoke-static>
</flow>
----
+
The Mule operation receives the `car` map whose key and values are mapped to a
`Car` class instance. This map is transformed into an instance before the invocation of

> the Java method. Without this improvement, the invocation would fail stating that the arguments

Reviewer: ```suggestion
the Java method. Without this improvement, the invocation would fail, stating that the arguments
```

passed to the Java method did not match its signature.

== Collections and Maps

> The Java Module checks each element of a Collection and each entry that a Map uses

Reviewer: ```suggestion
The Java module checks each element of a collection and each entry that a map uses
```
GLOBALLY lowercase these terms; they are not proper nouns, and we are not directly quoting the UI in context. I will stop marking.

Reviewer: Do we need to specify what it checks for? matching?

in the Java generic declared by the method signature. If an element fails this check, the whole

> collection or map has to be recreated and repopulated with the transformed elements.

Reviewer: ```suggestion
collection or map has to be re-created and repopulated with the transformed elements.
```
"re-created" is one exception to our style of not using hyphens after prefixes, because "recreated" has another meaning. GLOBAL.

> [WARNING]

Reviewer: Is this the type of notation appropriate for the content? See https://confluence.internal.salesforce.com/display/public/DOCTEAM/note.

====
The Java method is called with another instance of the collection or map 

> if a transformation is needed (not the same as the one passed in the Mule flow). This

Reviewer: Do we mean something like "If transformation is necessary, the Jave method is called using a collection or map instance that is different from the one passed in the Mule flow"?

Author: Checked with the developer and the audience will understand this well enough, plus trying to come up with different wording proved impossible.

</flow>
----

> The `numbers` list in the flow has two String objects (`"5"` and `"8"`), which the Module transforms into integers.

Reviewer: Per search results, lowercase "string objects":

Generally, String is a sequence of characters. But in Java, string is an object that represents a sequence of characters. The java.lang.String class is used to create a string object.

https://www.google.com/search?q=string+objects+in+java&oq=string+objects&aqs=chrome.2.69i57j0l5.7310j1j7&sourceid=chrome&ie=UTF-8

== Transform a Map Example

> A `Map` for the operation is transformed and

Reviewer: Ditto: a map vs. the `Map` value (or another common noun object to follow the init. cap. name)

> a new `Map` with the transformed keys and values is passed to the Java method.

Reviewer: ditto

</flow>
----

> This `Map` contains values that do not comply with the generics imposed in the Java

Reviewer: ditto
